### PR TITLE
Full WebAssembly toolchain

### PR DESCRIPTION
**Pull request by Absolucy**
This is a *full WebAssembly toolchain* for Procursus.  
It includes:
 - `build-essential-wasi`: A wasm32-wasi SDK/sysroot.
 - `binaryen`: A WebAssembly optimizer
 - `wabt`: Various WebAssembly tools
 - `wasmer`: A WebAssembly runtime (+ C API in `libwasmer`)
 	- Note: the Cranelift backend is used, rather than the LLVM backend. If anyone wants the LLVM backend, I could add support for it (just add `llvm` to the --features flag and add some environmental variable). tbh I just don't feel like running `make llvm` while away from home.
  - `wapm`: like npm but for WebAssembly

In addition, `llvm` was updated to compile with support for WebAssembly.

### CODE REVIEW

**Comment by Absolucy** [comment]
Note: wasmer won't build on iOS currently.

File: makefiles/wasmer.mk [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+wasmer-package: wasmer-stage
+	# wasmer.mk Package Structure
+	rm -rf $(BUILD_DIST)/wasmer $(BUILD_DIST)/libwasmer{,-dev}
+	mkdir -p $(BUILD_DIST)/wasmer $(BUILD_DIST)/libwasmer{,-dev}/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)
+
+	# wasmer.mk Prep wasmer
+	cp -a $(BUILD_STAGE)/wasmer $(BUILD_DIST)
+
+	# wasmer.mk Prep libwasmer
+	$(INSTALL) -Dm755 $(BUILD_STAGE)/libwasmer/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/libwasmer.dylib $(BUILD_DIST)/libwasmer/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/libwasmer.dylib
```
Comment (TheRealKeto) [inline makefiles/wasmer.mk:53]: Just copy to build_dist, instead. 

```suggestion
	cp -a $(BUILD_STAGE)/libwasmer/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/libwasmer.dylib $(BUILD_DIST)/libwasmer/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib/libwasmer.dylib
```

File: makefiles/wasmer.mk [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		--release \
+		--target=$(RUST_TARGET) \
+		--features="cranelift" \
+		--bin wasmer
+	unset CFLAGS && $(DEFAULT_RUST_FLAGS) cargo build \
+		--manifest-path="$(BUILD_WORK)/wasmer/lib/c-api/Cargo.toml" \
+		--release \
+		--target=$(RUST_TARGET) \
+		--features="cranelift,dylib,staticlib"
+	mkdir -p $(BUILD_STAGE)/libwasmer/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/lib
```
Comment (TheRealKeto) [inline makefiles/wasmer.mk:29]: You don't need this for the same reason line 31 works without making a bin folder.

```suggestion

```

File: makefiles/wasmer.mk [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	mkdir -p $(BUILD_STAGE)/libwasmer-dev/$(MEMO_PREFIX)$(MEMO_SUB_PREFIX)/{include,lib/pkgconfig}
```
Comment (TheRealKeto) [inline makefiles/wasmer.mk:30]: I suggest moving these to your setup stage instead (where GITHUB_ARCHIVE and EXTRACT_TAR goes)

File: build_info/wasmer.control [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Package: wasmer
+Version: @DEB_WASMER_V@
+Architecture: @DEB_ARCH@
+Maintainer: @DEB_MAINTAINER@
+Section: Scripting
+Priority: optional
+Homepage: https://wasmer.io/
+Description: An open-source runtime for executing WebAssembly.
+Tag: purpose::console, role::hacker
```
Comment (TheRealKeto) [inline build_info/wasmer.control:9]: We don't need this anymore.

```suggestion

```

File: build_info/wabt.control [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Package: wabt
+Version: @DEB_WABT_V@
+Architecture: @DEB_ARCH@
+Maintainer: @DEB_MAINTAINER@
+Section: Development
+Priority: optional
+Description: A suite of tools for WebAssembly.
+Homepage: https://webassembly.org
+Name: Web Assembly Binary Toolkit
+Tag: purpose::console, role::developer
```
Comment (TheRealKeto) [inline build_info/wabt.control:10]: We don't need this anymore, part 2

```suggestion

```

File: build_info/binaryen.control [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Package: binaryen
+Version: @DEB_BINARYEN_V@
+Architecture: @DEB_ARCH@
+Maintainer: @DEB_MAINTAINER@
+Section: Development
+Priority: optional
+Description: Binaryen is a compiler and toolchain infrastructure library for WebAssembly, written in C++. 
+ It aims to make compiling to WebAssembly easy, fast, and effective.
+Homepage: https://webassembly.org
+Tag: purpose::console, role::developer
```
Comment (TheRealKeto) [inline build_info/binaryen.control:10]: We don't need this (part 3, better description)

```suggestion
Description: compiler and toolchain infrastructure for WebAssembly.
 Binaryen is a compiler and toolchain infrastructure library for WebAssembly,
 written in C++. It aims to make compiling to WebAssembly easy, fast, and
 effective.
Homepage: https://webassembly.org
```

File: makefiles/wabt.mk [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	$(call GIT_CLONE,https://github.com/WebAssembly/wabt.git,$(WABT_VERSION),wabt)
+	mkdir -p $(BUILD_WORK)/wabt/build
+
+ifneq ($(wildcard $(BUILD_WORK)/wabt/.build_complete),)
+wabt:
+	@echo "Using previously built wabt."
+else
+wabt: wabt-setup
+	cd $(BUILD_WORK)/wabt/build && cmake \
+		-DBUILD_TESTS=OFF $(DEFAULT_CMAKE_FLAGS) \
```
Comment (TheRealKeto) [inline makefiles/wabt.mk:19]: I think you forgot to do this...

```suggestion
		$(DEFAULT_CMAKE_FLAGS) \
		-DBUILD_TESTS=OFF \
```

**Review by TheRealKeto — CHANGES_REQUESTED** [review]
QOL changes, only.

**Comment by Absolucy** [comment]
alright, this is ready. i made `wasmer` only available on macOS, and added `wapm`.

**Comment by Absolucy** [comment]
this is still ready...

**Comment by Absolucy** [comment]
oh, ok. take your time, sorry for bothering you about this /genuine

**Comment by Absolucy** [comment]
Sorry I uh, tend to close stuff whenever I'm in a bad mood.

**Pull request closed without merging**